File: servers/tides/src/tools/tide-tools.ts
import { z } from "zod";
import type { MCPTool } from "@packages/mcp-core/types.ts";
import { tideStorage } from "../storage/index.ts";

// Schemas for tidal workflow management
export const CreateTideInputSchema = z.object({
  name: z.string().describe("Name of the tidal workflow"),
  flow_type: z.enum(["daily", "weekly", "project", "seasonal"]).describe(
    "Type of tidal flow",
  ),
  description: z.string().optional().describe("Description of the workflow"),
});

export const CreateTideOutputSchema = z.object({
  success: z.boolean(),
  tide_id: z.string(),
  name: z.string(),
  flow_type: z.string(),
  created_at: z.string(),
  next_flow: z.string().optional(),
});

export const ListTidesInputSchema = z.object({
  flow_type: z.string().optional().describe("Filter by flow type"),
  active_only: z.boolean().optional().describe("Show only active tides"),
});

export const ListTidesOutputSchema = z.object({
  tides: z.array(z.object({
    id: z.string(),
    name: z.string(),
    flow_type: z.string(),
    status: z.enum(["active", "paused", "completed"]),
    created_at: z.string(),
    last_flow: z.string().optional(),
    next_flow: z.string().optional(),
  })),
  total: z.number(),
});

export const FlowTideInputSchema = z.object({
  tide_id: z.string().describe("ID of the tide to flow"),
  intensity: z.enum(["gentle", "moderate", "strong"]).optional().describe(
    "Flow intensity",
  ),
  duration: z.number().optional().describe("Flow duration in minutes"),
});

export const FlowTideOutputSchema = z.object({
  success: z.boolean(),
  tide_id: z.string(),
  flow_started: z.string(),
  estimated_completion: z.string(),
  flow_guidance: z.string(),
  next_actions: z.array(z.string()),
});

// Tool definitions
export const tideTools: MCPTool[] = [
  {
    name: "create_tide",
    description: "Create a new tidal workflow for rhythmic productivity",
    inputSchema: CreateTideInputSchema,
    outputSchema: CreateTideOutputSchema,
  },
  {
    name: "list_tides",
    description: "List all tidal workflows with their current status",
    inputSchema: ListTidesInputSchema,
    outputSchema: ListTidesOutputSchema,
  },
  {
    name: "flow_tide",
    description: "Start a flow session for a specific tidal workflow",
    inputSchema: FlowTideInputSchema,
    outputSchema: FlowTideOutputSchema,
  },
];

// Tool handlers
export const tideHandlers = {
  create_tide: async (args: z.infer<typeof CreateTideInputSchema>) => {
    const { name, flow_type, description } = args;

    try {
      const tide = await tideStorage.createTide({
        name,
        flow_type,
        description
      });

      console.error(`Creating tide: ${name} (${flow_type})`);

      return {
        success: true,
        tide_id: tide.id,
        name: tide.name,
        flow_type: tide.flow_type,
        created_at: tide.created_at,
        next_flow: tide.next_flow,
      };
    } catch (error) {
      console.error("Failed to create tide:", error);
      return {
        success: false,
        tide_id: "",
        name,
        flow_type,
        created_at: new Date().toISOString(),
      };
    }
  },

  list_tides: async (args: z.infer<typeof ListTidesInputSchema>) => {
    const { flow_type, active_only = false } = args;

    try {
      const tides = await tideStorage.listTides({
        flow_type,
        active_only
      });

      return {
        tides: tides.map(tide => ({
          id: tide.id,
          name: tide.name,
          flow_type: tide.flow_type,
          status: tide.status,
          created_at: tide.created_at,
          last_flow: tide.last_flow,
          next_flow: tide.next_flow,
        })),
        total: tides.length,
      };
    } catch (error) {
      console.error("Failed to list tides:", error);
      return {
        tides: [],
        total: 0,
      };
    }
  },

  flow_tide: async (args: z.infer<typeof FlowTideInputSchema>) => {
    const { tide_id, intensity = "moderate", duration = 25 } = args;

    try {
      // Verify tide exists
      const tide = await tideStorage.getTide(tide_id);
      if (!tide) {
        return {
          success: false,
          tide_id,
          flow_started: "",
          estimated_completion: "",
          flow_guidance: "Tide not found",
          next_actions: [],
        };
      }

      const flow_started = new Date().toISOString();
      const estimated_completion = new Date(
        Date.now() + duration * 60 * 1000,
      ).toISOString();

      // Add flow to tide history
      await tideStorage.addFlowToTide(tide_id, {
        timestamp: flow_started,
        intensity,
        duration,
      });

      // Generate guidance based on intensity
      const guidanceMap = {
        gentle:
          "🌊 Begin with calm, steady focus. Let thoughts flow naturally without forcing. Take breaks as needed.",
        moderate:
          "🌊 Maintain focused attention with deliberate action. Balance effort with ease. Stay present to the work.",
        strong:
          "🌊 Dive deep with sustained concentration. Channel energy into meaningful progress. Push through resistance mindfully.",
      };

      const flow_guidance = guidanceMap[intensity];

      const next_actions = [
        "🎯 Set clear intention for this flow session",
        "⏰ Start timer and begin focused work",
        "🧘 Take mindful breaks if needed",
        "📝 Capture insights and progress",
        "🌊 Honor the natural rhythm of the work",
      ];

      console.error(
        `Starting flow session for tide: ${tide_id} (${intensity} intensity, ${duration}min)`,
      );

      return {
        success: true,
        tide_id,
        flow_started,
        estimated_completion,
        flow_guidance,
        next_actions,
      };
    } catch (error) {
      console.error("Failed to start flow:", error);
      return {
        success: false,
        tide_id,
        flow_started: "",
        estimated_completion: "",
        flow_guidance: "Failed to start flow session",
        next_actions: [],
      };
    }
  },
};

File: servers/tides/src/tide_tools.py
"""
Tidal workflow management tools
"""

import logging
from datetime import datetime, timedelta
from typing import Any, Literal

from core.types import MCPTool
from pydantic import BaseModel, Field

from .tide_storage import CreateTideInput, FlowEntry, ListTidesFilter, TideStorage

logger = logging.getLogger(__name__)


# Initialize storage (use mounted volume in Docker or local directory for development)
import os

storage_path = os.getenv("TIDES_STORAGE_PATH", "./tides_data")
tide_storage = TideStorage(storage_path)


# Schema definitions
class CreateTideInputSchema(BaseModel):
    """Schema for creating a tide"""

    name: str = Field(description="Name of the tidal workflow")
    flow_type: Literal["daily", "weekly", "project", "seasonal"] = Field(
        description="Type of tidal flow"
    )
    description: str | None = Field(
        default=None, description="Description of the workflow"
    )


class CreateTideOutputSchema(BaseModel):
    """Schema for create tide output"""

    success: bool
    tide_id: str
    name: str
    flow_type: str
    created_at: str
    next_flow: str | None = None


class TideSummary(BaseModel):
    """Summary of a tide for listing"""

    id: str
    name: str
    flow_type: str
    status: Literal["active", "paused", "completed"]
    created_at: str
    last_flow: str | None = None
    next_flow: str | None = None


class ListTidesInputSchema(BaseModel):
    """Schema for listing tides"""

    flow_type: str | None = Field(default=None, description="Filter by flow type")
    active_only: bool | None = Field(default=None, description="Show only active tides")


class ListTidesOutputSchema(BaseModel):
    """Schema for list tides output"""

    tides: list[TideSummary]
    total: int


class FlowTideInputSchema(BaseModel):
    """Schema for flowing a tide"""

    tide_id: str = Field(description="ID of the tide to flow")
    intensity: Literal["gentle", "moderate", "strong"] | None = Field(
        default="moderate", description="Flow intensity"
    )
    duration: int | None = Field(default=25, description="Flow duration in minutes")


class FlowTideOutputSchema(BaseModel):
    """Schema for flow tide output"""

    success: bool
    tide_id: str
    flow_started: str
    estimated_completion: str
    flow_guidance: str
    next_actions: list[str]


# Tool definitions
tide_tools: list[MCPTool] = [
    MCPTool(
        name="create_tide",
        description="Create a new tidal workflow for rhythmic productivity",
        input_schema=CreateTideInputSchema.model_json_schema(),
        output_schema=CreateTideOutputSchema.model_json_schema(),
    ),
    MCPTool(
        name="list_tides",
        description="List all tidal workflows with their current status",
        input_schema=ListTidesInputSchema.model_json_schema(),
        output_schema=ListTidesOutputSchema.model_json_schema(),
    ),
    MCPTool(
        name="flow_tide",
        description="Start a flow session for a specific tidal workflow",
        input_schema=FlowTideInputSchema.model_json_schema(),
        output_schema=FlowTideOutputSchema.model_json_schema(),
    ),
]


# Tool handlers
async def create_tide_handler(args: dict[str, Any]) -> dict[str, Any]:
    """Handle tide creation"""
    validated_args = CreateTideInputSchema(**args)

    try:
        input_data = CreateTideInput(
            name=validated_args.name,
            flow_type=validated_args.flow_type,
            description=validated_args.description,
        )

        tide = await tide_storage.create_tide(input_data)

        logger.info(
            f"Creating tide: {validated_args.name} ({validated_args.flow_type})"
        )

        return CreateTideOutputSchema(
            success=True,
            tide_id=tide.id,
            name=tide.name,
            flow_type=tide.flow_type,
            created_at=tide.created_at,
            next_flow=tide.next_flow,
        ).model_dump()

    except Exception as error:
        logger.error(f"Failed to create tide: {error}")
        return CreateTideOutputSchema(
            success=False,
            tide_id="",
            name=validated_args.name,
            flow_type=validated_args.flow_type,
            created_at=datetime.now().isoformat(),
        ).model_dump()


async def list_tides_handler(args: dict[str, Any]) -> dict[str, Any]:
    """Handle listing tides"""
    validated_args = ListTidesInputSchema(**args)

    try:
        filter_data = ListTidesFilter(
            flow_type=validated_args.flow_type, active_only=validated_args.active_only
        )

        tides = await tide_storage.list_tides(filter_data)

        tide_summaries = [
            TideSummary(
                id=tide.id,
                name=tide.name,
                flow_type=tide.flow_type,
                status=tide.status,
                created_at=tide.created_at,
                last_flow=tide.last_flow,
                next_flow=tide.next_flow,
            )
            for tide in tides
        ]

        return ListTidesOutputSchema(
            tides=tide_summaries, total=len(tide_summaries)
        ).model_dump()

    except Exception as error:
        logger.error(f"Failed to list tides: {error}")
        return ListTidesOutputSchema(tides=[], total=0).model_dump()


async def flow_tide_handler(args: dict[str, Any]) -> dict[str, Any]:
    """Handle starting a flow session"""
    validated_args = FlowTideInputSchema(**args)

    try:
        # Verify tide exists
        tide = await tide_storage.get_tide(validated_args.tide_id)
        if not tide:
            return FlowTideOutputSchema(
                success=False,
                tide_id=validated_args.tide_id,
                flow_started="",
                estimated_completion="",
                flow_guidance="Tide not found",
                next_actions=[],
            ).model_dump()

        # Ensure we have valid values (Pydantic should provide defaults)
        intensity = validated_args.intensity or "moderate"
        duration = validated_args.duration or 25

        flow_started = datetime.now().isoformat()
        estimated_completion = (
            datetime.now() + timedelta(minutes=float(duration))
        ).isoformat()

        # Add flow to tide history
        flow_entry = FlowEntry(
            timestamp=flow_started,
            intensity=intensity,
            duration=duration,
        )

        await tide_storage.add_flow_to_tide(validated_args.tide_id, flow_entry)

        # Generate guidance based on intensity
        guidance_map = {
            "gentle": "🌊 Begin with calm, steady focus. Let thoughts flow naturally without forcing. Take breaks as needed.",
            "moderate": "🌊 Maintain focused attention with deliberate action. Balance effort with ease. Stay present to the work.",
            "strong": "🌊 Dive deep with sustained concentration. Channel energy into meaningful progress. Push through resistance mindfully.",
        }

        flow_guidance = guidance_map[intensity]

        next_actions = [
            "🎯 Set clear intention for this flow session",
            "⏰ Start timer and begin focused work",
            "🧘 Take mindful breaks if needed",
            "📝 Capture insights and progress",
            "🌊 Honor the natural rhythm of the work",
        ]

        logger.info(
            f"Starting flow session for tide: {validated_args.tide_id} "
            f"({validated_args.intensity} intensity, {validated_args.duration}min)"
        )

        return FlowTideOutputSchema(
            success=True,
            tide_id=validated_args.tide_id,
            flow_started=flow_started,
            estimated_completion=estimated_completion,
            flow_guidance=flow_guidance,
            next_actions=next_actions,
        ).model_dump()

    except Exception as error:
        logger.error(f"Failed to start flow: {error}")
        return FlowTideOutputSchema(
            success=False,
            tide_id=validated_args.tide_id,
            flow_started="",
            estimated_completion="",
            flow_guidance="Failed to start flow session",
            next_actions=[],
        ).model_dump()


# Handler mapping
tide_handlers = {
    "create_tide": create_tide_handler,
    "list_tides": list_tides_handler,
    "flow_tide": flow_tide_handler,
}
